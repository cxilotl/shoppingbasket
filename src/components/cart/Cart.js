import React from 'react';
import PropTypes from 'prop-types';
import cssStyles from "./Cart.module.scss";

const CartSubTotalItem = ({ id, name, price }) => {
  return (
    <li data-id={id} className={ cssStyles.subTotalListItem }><span>{ name }</span><span>{ `£${price.toFixed(2)}` }</span></li>
  );
};

CartSubTotalItem.propTypes = {
  id: PropTypes.string,
  name: PropTypes.string,
  price: PropTypes.number
};

const generateSubTotalItems = (items) => {
  let subTotalItems;
  if (items && Array.isArray(items) && items.length > 0) {
    subTotalItems = (
      <ul className={ cssStyles.subTotalList }>
        {
          items.map((item, index) => {
            const subTotalItemsPerProduct = [];
            if (item.quantity > 0) {
              for (let i = 0; i < item.quantity; i++) {
                subTotalItemsPerProduct.push(
                  <CartSubTotalItem
                    key={ `${item.name}-${index}-${i}` }
                    id={ `${item.name}-${i}` }
                    name={ item.name }
                    price={ item.price }
                  />
                );
              }
            }
            return subTotalItemsPerProduct;
          })
        }
      </ul>
    );
  }
  return subTotalItems;
};

const Cart = ({ items, subTotal, total }) => {
  const cartItems = generateSubTotalItems(items);
  return (
    <>
      <h2 className={ cssStyles.title }>Shopping Cart</h2>
      <div className={ cssStyles.cart }>
        <section data-testid="cart-sub-total" className={ cssStyles.subTotal }>
          { cartItems }
          <hr className={ cssStyles.subTotalSeparator } />
          <p className={ cssStyles.subTotalPrice }>
            <span>Sub-Total</span><span>{ `£${subTotal.toFixed(2)}` }</span>
          </p>
        </section>
        <hr className={ cssStyles.totalSeparator } />
        <div data-testid="cart-total" className={ cssStyles.total }>
          <span>Total</span><span>{ `£${total.toFixed(2)}` }</span>
        </div>
      </div>
    </>
  );
};

Cart.propTypes = {
  items: PropTypes.array,
  subTotal: PropTypes.number,
  total: PropTypes.number
};

export default Cart;