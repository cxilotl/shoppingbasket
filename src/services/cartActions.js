import { ADD_TO_CART, REMOVE_FROM_CART } from './actionTypes';

const addProduct = (product) => {
  return {
    type: ADD_TO_CART,
    payload: {
      product: product
    }
  }
};

const removeProduct = (product) => {
  return {
    type: REMOVE_FROM_CART,
    payload: {
      product: product
    }
  }
};

export {
  addProduct,
  removeProduct
};