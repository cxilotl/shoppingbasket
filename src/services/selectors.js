const getProducts = (state) => {
  return state.products;
};

const getProduct = (state, name) => {
  return state.products.find((product) => {
    return product.name === name;
  });
};

const isCartEmpty = (state) => {
  return (state.cart && state.cart.length === 0);
};

const getProductFromCart = (state, name) => {
  return state.cart.find((product) => {
    return product.name === name;
  });
};

const getTotalToPay = (state) => {
  return state.cart.reduce((total, productInCart) => {
    const productPricePerUnit = getProduct(state, productInCart.name);
    const totalPriceOfProduct = productPricePerUnit.price * productInCart.quantity;
    return total + totalPriceOfProduct;
  }, 0);
};

const getCartItems = (state) => {
  return state.cart.map((productInCart) => {
    return Object.assign({}, productInCart, {
      price: getProduct(state, productInCart.name).price
    });
  });
};

export {
  getProducts,
  getProduct,
  isCartEmpty,
  getProductFromCart,
  getTotalToPay,
  getCartItems
};