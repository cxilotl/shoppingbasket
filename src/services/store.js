import { createStore } from "redux";
import reducer from './reducers';
import products from '../products';

const store = createStore(
  reducer,
  {
    products: products,
    cart: []
  },
  window.__REDUX_DEVTOOLS_EXTENSION__ && window.__REDUX_DEVTOOLS_EXTENSION__() // for debugging
);

export { store };