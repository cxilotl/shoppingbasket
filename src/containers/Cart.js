import { connect } from 'react-redux';
import Cart from "../components/cart/Cart";
import { getTotalToPay, getCartItems } from "../services/selectors";

const mapStateToProps = (state) => {
  return {
    items: getCartItems(state),
    subTotal: getTotalToPay(state),
    total: getTotalToPay(state)
  };
};

const mapDispatchToProps = null;

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Cart);