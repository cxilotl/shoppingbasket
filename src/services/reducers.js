import { ADD_TO_CART, REMOVE_FROM_CART } from './actionTypes';
import { isCartEmpty, getProductFromCart } from "./selectors";

const handleAddProductToCart = (state, payload) => {
  if (typeof payload.product !== 'undefined') {
    if (!isCartEmpty(state)) {
      if (getProductFromCart(state, payload.product)) {
        return {
          ...state,
          cart: state.cart.map((productItem) => {
            if (productItem.name === payload.product) {
              return {
                name: productItem.name,
                quantity: productItem.quantity + 1
              }
            } else {
              return productItem;
            }
          })
        };
      } else {
        return {
          ...state,
          cart: [
            ...state.cart,
            {
              name: payload.product,
              quantity: 1
            }
          ]
        };
      }
    } else {
      return {
        ...state,
        cart: [{
          name: payload.product,
          quantity: 1
        }]
      };
    }
  }
  return state;
};

const handleRemoveProductFromCart = (state, payload) => {
  if (typeof payload.product !== 'undefined' && !isCartEmpty(state) && getProductFromCart(state, payload.product)) {
    const existingProduct = getProductFromCart(state, payload.product);
    if (existingProduct.quantity > 1) {
      return {
        ...state,
        cart: state.cart.map((productItem) => {
          if (productItem.name === payload.product) {
            return {
              name: productItem.name,
              quantity: productItem.quantity - 1
            }
          } else {
            return productItem;
          }
        })
      }
    } else {
      return {
        ...state,
        cart: state.cart.filter((product) => {
          return product.name !== payload.product;
        })
      };
    }
  }
  return state;
};

const initialState = {
  products: [],
  cart: []
};

const reducer = (state = initialState, action = {}) => {
  switch (action.type) {
    case ADD_TO_CART:
      return handleAddProductToCart(state, action.payload);
    case REMOVE_FROM_CART:
      return handleRemoveProductFromCart(state, action.payload);
    default:
      return state;
  }
};

export default reducer;