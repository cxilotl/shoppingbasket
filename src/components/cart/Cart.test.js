import React from 'react';
import { Provider } from "react-redux";
import renderer from 'react-test-renderer';
import Cart from './Cart';

import { store as appStore } from "../../services/store";

describe('Cart', () => {
  it('Snapshot', () => {
    const items = [
      { name: 'Beans', price: 0.5, quantity: 3 },
      { name: 'Coke', price: 0.7, quantity: 1 }
    ];
    const subTotal = 2.2;
    const total = 2.2;
    const component = renderer.create(
      <Provider store={ appStore }>
        <Cart
          items={ items }
          subTotal={ subTotal }
          total={ total }
        />
      </Provider>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});