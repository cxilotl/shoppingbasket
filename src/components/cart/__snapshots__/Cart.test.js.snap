// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cart Snapshot 1`] = `
<div
  className="cart"
>
  <section
    className="subTotal"
    data-testid="cart-sub-total"
  >
    <ul
      className="subTotalList"
    >
      <li
        className="subTotalListItem"
        data-id="Beans-0"
      >
        <span>
          Beans
        </span>
        <span>
          £0.50
        </span>
      </li>
      <li
        className="subTotalListItem"
        data-id="Beans-1"
      >
        <span>
          Beans
        </span>
        <span>
          £0.50
        </span>
      </li>
      <li
        className="subTotalListItem"
        data-id="Beans-2"
      >
        <span>
          Beans
        </span>
        <span>
          £0.50
        </span>
      </li>
      <li
        className="subTotalListItem"
        data-id="Coke-0"
      >
        <span>
          Coke
        </span>
        <span>
          £0.70
        </span>
      </li>
    </ul>
    <hr
      className="subTotalSeparator"
    />
    <p
      className="subTotalPrice"
    >
      <span>
        Sub-Total
      </span>
      <span>
        £2.2
      </span>
    </p>
  </section>
  <hr
    className="totalSeparator"
  />
  <div
    className="total"
    data-testid="cart-total"
  >
    <span>
      Total
    </span>
    <span>
      £2.2
    </span>
  </div>
</div>
`;
